Refactor "Check for updates" dialog

Johno-ACSLive opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #6738


## Proposed changes

Updated button text from "Download Now" to "Update Now".

Updated Update Now button / event names.

Moved Update Now button position to where the Close button was located.

Added a new link called "Direct Download" which restores original functionality of launching default browser to download MSI.


## Screenshots <!-- Remove this section if PR does not change UI -->

### After

<!-- TODO -->
![image](https://user-images.githubusercontent.com/45373169/59495806-74c7b700-8ed3-11e9-94cd-d370cbde4245.png)
----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.Designer.cs:
```diff
@@ -30,24 +30,13 @@ private void InitializeComponent()
-            this.closeButton.TabIndex = 0;
-            this.closeButton.Text = "Close";
-            this.closeButton.UseVisualStyleBackColor = true;
-            this.closeButton.Click += new System.EventHandler(this.CloseButtonClick);
```
The handler must be removed

Johno-ACSLive replied:
Hmm thought i removed those, good catch!

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.cs:
```diff
@@ -177,6 +178,17 @@ private async void btnDownloadNow_Click(object sender, EventArgs e)
+            }
+            catch (Win32Exception)
+            {
+            }
```
* What sort of exceptions are we expecting?
* By swallowing exceptions we're creating a new problem - something isn't working and we don't know the cause. It is better to throw and let the NBug handler create a report.

Johno-ACSLive replied:
I just restored the original code which didn't have anything. I'm not entirely sure what exceptions we would expect when we tell the OS to launch a URL via the Default browser.

Rather than catch a Win32 exception, perhaps we catch all exceptions and we can send it via NBug. Are you happy with this approach?

I haven't used NBug but i assume it is similar to Microsoft AppCentre.

RussKie replied:
No need to `try..catch` - an exception will be trapped by NBug via global handlers

Johno-ACSLive replied:
Great, I'll remove the try..catch in that case

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.Designer.cs:
```diff
@@ -80,27 +69,40 @@ private void InitializeComponent()
             // 
             // FormUpdates
             // 
             this.AutoScaleDimensions = new System.Drawing.SizeF(120F, 120F);
```
It looks like the last changes were applied in a scaled Windows.
We need to fix it and run all designers at 96dpi.

If you're not running Windows at 100% scale, please start your VS with `/noscale` switch.

Johno-ACSLive replied:
Ahh yes, I tried to keep everything the same. Looks like this slipped through as i made the changed on my laptop which has a high DPI screen. I'll correct this!

Johno-ACSLive replied:
Looks like you corrected this with your UI tweak 👍

RussKie commented:
I've made a little tweak to the UI:
![image](https://user-images.githubusercontent.com/4403806/59559157-e1ae8e80-9044-11e9-84c6-f4bd6bb8a317.png)

RussKie commented:
Also we need to apply keyboard accelerators to the controls as well.

Johno-ACSLive commented:
> Also we need to apply keyboard accelerators to the controls as well.

Do you mean apply keyboard shortcuts (e.g. Ctrl + U to Update Now)?

RussKie commented:
> > Also we need to apply keyboard accelerators to the controls as well.
> 
> Do you mean apply keyboard shortcuts (e.g. Ctrl + U to Update Now)?

Yes.

Open for a discussion, but I propose "&Show chanlog", "Up&date now" and "Direct do&wnload".
This way you can invoke all of these with one hand, without needing to reach across the keyboard...

Johno-ACSLive commented:
> > > Also we need to apply keyboard accelerators to the controls as well.
> > 
> > 
> > Do you mean apply keyboard shortcuts (e.g. Ctrl + U to Update Now)?
> 
> Yes.
> 
> Open for a discussion, but I propose "&Show chanlog", "Up&date now" and "Direct do&wnload".
> This way you can invoke all of these with one hand, without needing to reach across the keyboard...

Good suggestion! I've always preferred the shortcut to reflect more the action and also avoid common shortcuts e.g. Ctrl + C etc. but i guess given the context it may not really matter.

We could also use:
- "Show Change&log"
- "&Update Now"
- "&Direct Download"

Happy to go either way, if we are not happy with what goes in we can always change it later.

gerhardol commented:
> We could also use:
L, D, U

What I prefer.
Keyboard shortcuts are not important for me though.

RussKie commented:
Fine, you both win :\

On Mon, 17 Jun 2019 at 22:24, Gerhard Olsson <notifications@github.com>
wrote:

> We could also use:
> L, D, U
>
> What I prefer.
> Keyboard shortcuts are not important for me though.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6801?email_source=notifications&email_token=ABBTEXSV7HGGGLED3FXEDBDP257AVA5CNFSM4HYFX4IKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX276EY#issuecomment-502660883>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXXKX562OXNYZBE6H4TP257AVANCNFSM4HYFX4IA>
> .
>

Johno-ACSLive commented:
OK, so the mnemonic shortcuts (adding the & in front of the character that will be used for the HotKey) of Buttons are using Alt.

Doesn't work for labels or links, so I have worked around this limitation.

The UI tweak also introduced a UI defect when clicking Update Now in that it covered up the progress bar. I've added a line to hide the link when the Update Now button is clicked (and also the relevant HotKey is used).

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.cs:
```diff
@@ -177,6 +174,45 @@ private async void btnDownloadNow_Click(object sender, EventArgs e)
+            }
+        }
+
+        protected override bool ProcessCmdKey(ref Message msg, Keys keyData)
```
[nit] Please move before all private methods

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.cs:
```diff
@@ -177,6 +174,45 @@ private async void btnDownloadNow_Click(object sender, EventArgs e)
+            LaunchURL(LaunchType.DirectDownload);
+        }
+
+        private void LaunchURL(LaunchType launchType)
```
```suggestion
        private void LaunchUrl(LaunchType launchType)
```

RussKie replied:
[nit] Please move above event handlers

RussKie commented on GitUI/CommandsDialogs/BrowseDialog/FormUpdates.cs:
```diff
@@ -177,6 +174,45 @@ private async void btnDownloadNow_Click(object sender, EventArgs e)
+        }
+    }
+
+    public enum LaunchType
```
Does is have to be `public`?

Johno-ACSLive replied:
I just cloned the ReleaseType enum. Declaring as private "Error	CS1527	Elements defined in a namespace cannot be explicitly declared as private, protected, protected internal, or private protected".

Internal works, are you happy for me to set this to internal?

Johno-ACSLive replied:
Set to internal on current commit

RussKie replied:
That's fine, thanks.

RussKie commented:
One more thing, please update translations as described here: https://github.com/gitextensions/gitextensions/wiki/Translations#how-to-update-english-translations

The latest script is available in e7688f6fd82614979d964cf69233b951ad521970

Johno-ACSLive commented:
> One more thing, please update translations as described here: https://github.com/gitextensions/gitextensions/wiki/Translations#how-to-update-english-translations
> 
> The latest script is available in [e7688f6](https://github.com/gitextensions/gitextensions/commit/e7688f6fd82614979d964cf69233b951ad521970)

Not a problem, will go through that now

Johno-ACSLive commented:
Did you want the Translation to be included in this PR or a new PR?

RussKie commented:
> Did you want the Translation to be included in this PR or a new PR?

As part of this PR, thanks

Johno-ACSLive commented:
> > Did you want the Translation to be included in this PR or a new PR?
> 
> As part of this PR, thanks

Done

RussKie commented:
When you think you're done, take off the WIP and squash

RussKie commented:
When you think you're done, please remove WIP and tidy up the PR (squash what's necessary etc) and let us know.

Johno-ACSLive commented:
> When you think you're done, take off the WIP and squash

I believe everything is now addressed and translations have been updated

Johno-ACSLive commented:
> and squash

Do you mean rebase and merge all 10 commits in this PR to a single commit?

RussKie commented:
> Do you mean rebase and merge all 10 commits in this PR to a single commit?

Yes. 
You can find more details here - https://github.com/gitextensions/gitextensions/wiki#committing

Johno-ACSLive commented:
> > Do you mean rebase and merge all 10 commits in this PR to a single commit?
> 
> Yes.
> You can find more details here - https://github.com/gitextensions/gitextensions/wiki#committing

I think it's now done, 1 commit with all changes. My first squash, hopefully I did it correctly.

RussKie commented:
> My first squash, hopefully I did it correctly.

Congratulations! 🎉

RussKie commented:
Thank you

Johno-ACSLive commented:
> > My first squash, hopefully I did it correctly.
> 
> Congratulations! 🎉

Thank you! 😁
